modules/twinklexfd.js

```diff
@@ -1313,18 +1300,38 @@ Twinkle.xfd.callbacks = {
 
 
 	rfd: {
-		// This is a callback from an API request, which gets the target of the redirect
-		findTargetCallback: function(apiobj) {
-			var xmlDoc = apiobj.responseXML;
-			var target = $(xmlDoc).find('redirects r').first().attr('to');
-			if( !target ) {
-				apiobj.statelem.error( "This page is currently not a redirect, aborting" );
-				return;
+		// This gets called both on submit and preview to determine the redirect target
+		findTarget: function(params, callback) {
+			if (document.getElementById("softredirect")) {
+				// For soft redirects, skip straight to the callback
+				params.target = document.getElementById("softredirect").textContent.replace(/^:+/, "");
+				callback(params);
+			} else {
+				// Find current target of redirect
+				var query = {
+					'action': 'query',
+					'titles': mw.config.get('wgPageName'),
+					'redirects': true
+				};
+				var wikipedia_api = new Morebits.wiki.api( "Finding target of redirect", query, Twinkle.xfd.callbacks.rfd.findTargetCallback(callback) );
+				wikipedia_api.params = params;
+				wikipedia_api.post();
 			}
-			apiobj.params.target = target;
-			var section = $(xmlDoc).find('redirects r').first().attr('tofragment');
-			apiobj.params.section = section;
-			Twinkle.xfd.callbacks.rfd.main(apiobj.params);
+		},
+		// This is a closure for the callback from the above API request, which gets the target of the redirect
+		findTargetCallback: function(callback) {
+			return function(apiobj) {
+				var xmlDoc = apiobj.responseXML;
+				var target = $(xmlDoc).find('redirects r').first().attr('to');
+				if( !target ) {
+					apiobj.statelem.error( "This page is currently not a redirect, aborting" );
```

Comment:
The preview will fail silently (except for the console message) here if the page isn't a redirect.  Certainly an improvement over the current situation, but I'm not entirely sure how to treat the preview (don't quit) differently than the submission (quit with error overlay), or if it's even worth it.

Reply:
This also happens for the ffd check above, so it's not anything new.  We should handle these better in general, but I think this can go forward.